agent/pkg/proxy/controller/controller.go

```diff
@@ -0,0 +1,138 @@
+package controller
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubeedge/edgemesh/agent/pkg/common/informers"
+)
+
+type ProxyController interface {
+	Init()
+	GetSvcPorts(ip string) string
+	GetSvcIP(svcName string) string
+}
+
+type controller struct {
+	svcLister   listers.ServiceLister
+	svcInformer cache.SharedIndexInformer
+
+	sync.RWMutex
+	SvcPortsByIP map[string]string // key: clusterIP, value: SvcPorts
+	IPBySvc      map[string]string // key: svcName.svcNamespace, value: clusterIP
+}
+
+func New(ifm *informers.Manager) *controller {
```

Comment:
I think function New should return a singleton, if other package want to use the interface of controller, now have to import it in the package proxy, and new a proxy to call it.

Reply:
Adopt your suggestion